docs/src/main/mdoc/datatypes/statet.md

```diff
@@ -0,0 +1,244 @@
+error is returned.
+
+To simplify the logic, for each reservation we will just consider a
+single `LocalTime` starting at the beginning of the hour.
+
+Let's start with defining the type synonyms for the state of the
+program and the effect type:
+
+```scala mdoc:silent
+import cats.data.{StateT, NonEmptyList}
+import cats.implicits._
+import java.time.LocalTime
+
+type TRSState = Map[(Int, LocalTime), String]
+type EffectType[A] = Either[Throwable, A]
+```
+
+
+`TRSState` stands for *Table Reservation System State* and it will
+organize the reservations in a `Map` with:
+- `(Int, LocalTime)` as key, representing the table number and hour
+respectively.
+- `String` as value, representing the reservation name.
+
+As effect type, we will just use an `Either[Throwable, A]` because each
+booking insertion can fail.
+
+Now, we need to implement/define:
+- An initial state, that will be just an empty `Map`.
+- A custom `Throwable` to be used in case of an error.
+- The logic for the booking insertion. We can take advantage of the
+method `modifyF`.
+
+In addition, we can implement a simple function that will evaluate a
+`NonEmptyList` of reservations, inserting them one by one.
+
+```scala mdoc:silent
+object TableReservationSystem {
+
+  final case class AlreadyReservedTable(
+      tableNumber: Int,
+      time: LocalTime,
+      reservation: String
+  ) extends Throwable(
+        s"$reservation cannot be added because table number $tableNumber is already reserved for the $time"
+      )
+
+  val emptyReservationSystem: TRSState = Map.empty
+
+  def insertBooking(
+      tableNumber: Int,
+      time: LocalTime,
+      name: String
+  ): StateT[EffectType, TRSState, Unit] =
+    StateT.modifyF[EffectType, TRSState]((currentReservations: TRSState) =>
+      currentReservations
+        .get((tableNumber, time))
+        .fold[EffectType[TRSState]](Right(currentReservations + ((tableNumber, time) -> name)))(_ =>
+          Left(AlreadyReservedTable(tableNumber, time, name))
+        )
+    )
+
+  def evalBookings(
+      bookings: NonEmptyList[(Int, LocalTime, String)]
+  ): EffectType[TRSState] =
+    bookings
+      .map { case (tn, t, n) =>
+        TableReservationSystem.insertBooking(tn, t, n)
+      }
+      .reduceLeft(_ *> _)
+      .runS(TableReservationSystem.emptyReservationSystem)
+}
+```
+
+That's it, we can finally test the code above providing a simple
+example input and print out the result.
+
+```scala mdoc
+val bookings = NonEmptyList.of(
+  (1, LocalTime.parse("10:00:00"), "Cristiano Ronaldo"),
+  (2, LocalTime.parse("10:00:00"), "Lebron James"),
+  (1, LocalTime.parse("12:00:00"), "Tiger Woods"),
+  (2, LocalTime.parse("12:00:00"), "Roger Federer"),
+  (3, LocalTime.parse("13:00:00"), "Elizabeth Alexandra Mary"),
+  (1, LocalTime.parse("16:00:00"), "Chuck Norris"),
+  (2, LocalTime.parse("16:00:00"), "Kobe Bryant"),
+  (2, LocalTime.parse("18:00:00"), "Steven Seagal")
+)
+
+println(TableReservationSystem.evalBookings(bookings))
```

Comment:
We don't need `println` in mdoc :) if you write `TableReservationSystem.evalBookings(bookings)` it will automatically put it in the output